File: blooming_directed_graph_agents/filter_agents/csharp/Collection.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace LangGraph;

/// <summary>blooming_directed_graph_filter_agents – C# collection.</summary>
public static class Collection
{
    public static List<Node> FilterByReceiver(List<Node> nodes, string receiver)
    {
        if (receiver != "github" && receiver != "hpcc") return nodes;
        return nodes.Where(n => n.Owner != null && n.Repo != null).ToList();
    }

    public static List<Node> FilterByActionWhere(List<Node> nodes, string where)
    {
        if (where != "github" && where != "hpcc") return nodes;
        return nodes.Where(n =>
            (where == "github" && (n.Name?.Contains("github") == true)) ||
            (where == "hpcc" && (n.Name?.ToLowerInvariant().Contains("hpcc") == true))).ToList();
    }

    public static List<Node> FilterByActionClient(List<Node> nodes, string client)
    {
        if (client != "macbook" && client != "rockydesktop") return nodes;
        return nodes.Where(n =>
            (client == "macbook" && (n.Name?.Contains("owner") == true)) ||
            (client == "rockydesktop" && (n.Name?.Contains("quay") == true))).ToList();
    }

    public static List<Node> FilterByRepoName(List<Node> nodes, Regex pattern)
    {
        return nodes.Where(n => n.Name != null && pattern.IsMatch(n.Name)).ToList();
    }

    public class Node
    {
        public string? Name { get; set; }
        public string? Path { get; set; }
        public string? Owner { get; set; }
        public string? Repo { get; set; }
        public string? Slug { get; set; }
    }
}

File: blooming_directed_graph_agents/filter_agents/swift/Collection.swift
// blooming_directed_graph_filter_agents – Swift collection.
// Namespace: LangGraph (graph filter agents).
// Filter nodes by receiver, action_where, action_client, repo name.

import Foundation

/// LangGraph: graph node for filter agents.
public struct Node {
    public var name: String
    public var path: String
    public var owner: String?
    public var repo: String?
    public var slug: String
    public init(name: String, path: String, owner: String?, repo: String?, slug: String) {
        self.name = name
        self.path = path
        self.owner = owner
        self.repo = repo
        self.slug = slug
    }
}

/// LangGraph namespace: filter agents collection.
public enum BloomingDirectedGraphFilterAgents {
    public static func filterByReceiver(_ nodes: [Node], receiver: String) -> [Node] {
        guard receiver == "github" || receiver == "hpcc" else { return nodes }
        return nodes.filter { $0.owner != nil && $0.repo != nil }
    }

    public static func filterByActionWhere(_ nodes: [Node], where where_: String) -> [Node] {
        guard where_ == "github" || where_ == "hpcc" else { return nodes }
        return nodes.filter {
            where_ == "github" ? $0.name.contains("github") : $0.name.lowercased().contains("hpcc")
        }
    }

    public static func filterByActionClient(_ nodes: [Node], client: String) -> [Node] {
        guard client == "macbook" || client == "rockydesktop" else { return nodes }
        return nodes.filter {
            client == "macbook" ? $0.name.contains("owner") : $0.name.contains("quay")
        }
    }

    public static func filterByRepoName(_ nodes: [Node], pattern: String) -> [Node] {
        return nodes.filter { $0.name.contains(pattern) }
    }
}
